doc/designs/thread-api.md

```diff
@@ -0,0 +1,304 @@
+
+  - Some algorithms are designed to be run in parallel (Argon2);
+  - Some transports (e.g. QUIC, DTLS) may need to handle timer events
+    independently of application calls to OpenSSL.
+
+To support this end, OpenSSL can manage an internal thread pool. Tasks can be
+scheduled on the internal thread pool.
+
+There are three options given to an application which wants to use the thread
+pool functionality:
+
+  1. Use OpenSSL's default thread pool functionality.
+
+  2. Use OpenSSL's thread pool functionality with a custom thread spawner
+     method.
+
+  3. Provide threads manually to OpenSSL for use as part of its thread pool.
+
+Default Model
+-------------
+
+In the default model, OpenSSL creates and manages threads up to a maximum
+number of threads authorized by the application.
+
+The application enables thread pooling by calling the following function
+during its initialisation:
+
+```c
+/*
+ * Set the maximum number of threads to be used by the thread pool.
+ *
+ * If the argument is 0, thread pooling is disabled. OpenSSL will not create any
+ * threads and existing threads in the thread pool will be torn down.
+ *
+ * Returns 1 on success and 0 on failure. Returns failure if OpenSSL-managed
+ * thread pooling is not supported (for example, if it is not supported on the
+ * current platform, or because OpenSSL is not built with the necessary
+ * support).
+ */
+int OSSL_set_max_threads(uint32_t max_threads);
+
+/*
+ * Get the maximum number of threads currently allowed to be used by the
+ * thread pool. If thread pooling is disabled or not available, returns 0.
+ */
+uint32_t OSSL_get_max_threads(void);
+```
+
+The maximum thread count is a limit, not a target. Threads will not be spawned
+unless (and until) there is demand.
+
+Custom Thread Spawner Method
+----------------------------
+
+If an application wishes to control thread spawning, it may set a custom set of
+thread method functions. OpenSSL will use these to spawn and join threads.
+
+```c
+typedef struct ossl_thread_cookie_st OSSL_THREAD_COOKIE;
+
+typedef struct ossl_thread_method_st {
+    /*
+     * Spawns thread, calling cb with arg. An opaque cookie representing the
+     * thread is written to *thread_cookie.
+     *
+     * Returns 1 on success or 0 on failure.
+     */
+    int (*spawn_thread)(void (*cb)(void *arg), void *arg,
+                        OSSL_THREAD_COOKIE **thread_cookie);
+
+    /*
+     * Wait for a thread to exit.
+     */
+    void (*join_thread)(OSSL_THREAD_COOKIE *thread_cookie);
+} OSSL_THREAD_METHOD;
+
+/*
+ * Call OSSL_set_thread_method(&method, sizeof(method)) to set a new thread
+ * method. This must be called before OpenSSL creates any threads (e.g. before
+ * the call to `OSSL_set_max_threads`), else it fails.
+ *
+ * The length argument allows ABI compatibility to be preserved if methods are
+ * added to the OSSL_THREAD_METHOD structure in future.
+ */
+int OSSL_set_thread_method(OSSL_THREAD_METHOD *method, size_t len);
```

Comment:
Hmmm, this would be the very first time we've used the Windows length of structure as a versioning system.  Could we include a version byte at the start instead?  I much prefer a version byte to a random length -- the length doesn't let you add & remove stuff properly.

Reply:
I'm OK with this.